examples/dataflow-xml-pubsub-to-gcs/python/beamPubSubXml2Gcs.py

```diff
@@ -0,0 +1,207 @@
+            for element in parsedXml:
+                allTags.append(element.tag)
+                allTagsText.append(element.text)
+            yield {"tags": allTags, "text": allTagsText}
+        except Exception as e:
+            yield {"error": str(e), "raw_contents": message_body}
+
+
+class WriteToGCS(DoFn):
+    def __init__(self, output_path):
+        self.output_path = output_path
+
+    def process(self, key_value, window=DoFn.WindowParam):
+        """Write messages in a batch to Google Cloud Storage."""
+
+        ts_format = "%H:%M"
+        window_start = window.start.to_utc_datetime().strftime(ts_format)
+        window_end = window.end.to_utc_datetime().strftime(ts_format)
+        shard_id, parsed_payload = key_value
+        filename = "{0}streaming_data-{1}-{2}-{3}.txt".format(self.output_path,
+                                                              window_start,
+                                                              window_end,
+                                                              str(shard_id))
+
+        with io.gcsio.GcsIO().open(filename=filename, mode="w") as f:
+            f.write(f"{parsed_payload}\n".encode("utf-8"))
+
+
+def run(project_id,
+        input_topic,
+        gcs_path,
+        window_size,
+        num_shards,
+        runner,
+        region,
+        pipeline_args=None):
+    # Set `save_main_session` to True so DoFns can access globally imported
+    # modules.
+    input_topic = "projects/{0}/topics/{1}".format(project_id, input_topic)
+    if gcs_path[-1] == "/":
+        gcs_path = gcs_path[:-1]
+    output_path = "{0}/output/".format(gcs_path)
+    provided_args = {
+        "project": project_id,
+        "runner": runner,
+        "region": region,
+        "staging_location": "{0}/staging/".format(gcs_path),
+        "temp_location": "{0}/temp/".format(gcs_path),
+        "streaming": True,
+        "save_main_session": True
+        }
+
+    pipeline_options = PipelineOptions(
+        pipeline_args, **provided_args
+    )
+
+    with Pipeline(options=pipeline_options) as pipeline:
+        (
+            pipeline
+            # Because `timestamp_attribute` is unspecified in `ReadFromPubSub`,
+            # Beam binds the publish time returned by the Pub/Sub server for
+            # each message to the element's timestamp parameter, accessible via
+            # `DoFn.TimestampParam`.
+            # https://beam.apache.org/releases/pydoc/current/apache_beam.io.gcp.pubsub.html#apache_beam.io.gcp.pubsub.ReadFromPubSub
+            # https://cloud.google.com/pubsub/docs/stream-messages-dataflow#set_up_your_pubsub_project
+            | "Read from Pub/Sub" >> io.ReadFromPubSub(topic=input_topic)
+            | "Parse XML tags and attributes" >> ParDo(ParseXML())
+            | "Window into" >> GroupMessagesByFixedWindows(window_size,
+                                                           num_shards)
+            | "Write to GCS" >> ParDo(WriteToGCS(output_path))
```

Comment:
I would replace this part and use the built in [WriteToText](https://beam.apache.org/releases/pydoc/2.41.0/apache_beam.io.textio.html#apache_beam.io.textio.WriteToText) transform. With this transform you should be able to scale writes better than using file systems per pCollection item. 

You might need to add a step before for pre-processing / grouping of data depending on the structure of the data you have in the PCollection

Reply:
Addressed in latest commit, please review